src/proxy/http/settings.rs

```diff
@@ -10,95 +8,81 @@ use super::h1;
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub enum Settings {
     Http1 {
-        host: Host,
-        /// Whether the request wants to use HTTP/1.1's Upgrade mechanism.
+        /// Indicates whhether a new service must be created for each request.
+        stack_per_request: bool,
+         /// Whether the request wants to use HTTP/1.1's Upgrade mechanism.
         ///
         /// Since these cannot be translated into orig-proto, it must be
         /// tracked here so as to allow those with `is_h1_upgrade: true` to
         /// use an explicitly HTTP/1 service, instead of one that might
         /// utilize orig-proto.
         is_h1_upgrade: bool,
-        /// Whether or not the request URI was in absolute form.
+         /// Whether or not the request URI was in absolute form.
         ///
         /// This is used to configure Hyper's behaviour at the connection
         /// level, so it's necessary that requests with and without
         /// absolute URIs be bound to separate service stacks. It is also
         /// used to determine what URI normalization will be necessary.
         was_absolute_form: bool,
-    },
-    Http2
-}
-
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum Host {
-    Authority(uri::Authority),
-    NoAuthority,
+   },
+    Http2,
 }
 
 // ===== impl Settings =====
 
 impl Settings {
-    pub fn detect<B>(req: &http::Request<B>) -> Self {
+    pub fn from_request<B>(req: &http::Request<B>) -> Self {
         if req.version() == http::Version::HTTP_2 {
             return Settings::Http2;
         }
 
-        let was_absolute_form = super::h1::is_absolute_form(req.uri());
-        trace!(
-            "Settings::detect(); req.uri='{:?}'; was_absolute_form={:?};",
-            req.uri(), was_absolute_form
-        );
-        // If the request has an authority part, use that as the host part of
-        // the key for an HTTP/1.x request.
-        let host = Host::detect(req);
+        let has_name = req
+            .uri()
+            .authority_part()
+            .map(|_| true)
+            .or_else(|| {
+                req.headers()
+                    .get(HOST)
+                    .and_then(|h| h.to_str().ok())
+                    .map(|h| !h.is_empty())
+            })
+            .unwrap_or(false);
 
+        let was_absolute_form = super::h1::is_absolute_form(req.uri());
         let is_h1_upgrade = super::h1::wants_upgrade(req);
-
         Settings::Http1 {
-            host,
-            is_h1_upgrade,
             was_absolute_form,
+            is_h1_upgrade,
+            stack_per_request: !has_name,
```

Comment:
TIOLI: I think the logic here could be simplified so that we have to do fewer negations. Consider something like:
```rust
        let stack_per_request = req
            .uri()
            .authority_part()
            .map(|_| false)
            .or_else(|| {
                req.headers()
                    .get(HOST)
                    .and_then(|h| h.to_str().ok())
                    .map(|h| h.is_empty())
            })
            .unwrap_or(true);

         // ...

        Settings::Http1 {
            // ...
            stack_per_request,
        }
```

Reply:
I had this initially but I found it confusing.  I’ll see if I can land on something that reduces negations but is still readable (probably naming the temp Val `missing_host` or something)

Reply:
Okay, that's also fair. I just found the number of negations made it hard to think through. 

Maybe more intermediate `let`s (and/or some comments) would help?